### ISSUE TITLE
Jitted function cannot be pickled

### ISSUE THREAD

**Issue by twoertwein** [pytorch/pytorch#61210]
## 🚀 Feature
Support to pickle a jitted function (or at least throw a `TypeError` when using protocol 0 and 1).

## Motivation

Trying to pickle a jitted function either raises `TypeError: cannot pickle 'torch._C.ScriptFunction' object` when `protocol>1` or far worse when using `protocol=0` or `protocol=1` python 3.9.5 dies with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  instance allocation failed: new instance has no pybind11-registered base types
Aborted (core dumped)
```

Example:
```py
import pickle

import torch

@torch.jit.script
def fun(x: torch.Tensor) -> torch.Tensor:
    return x

pickle.dumps(fun, protocol=0)

```
## Pitch
Implement pickle support for all pickle protocols (or raise a `TypeError` when used with `protocl<2`).

## Additional context
I encountered this when using a jitted function with dask https://github.com/dask/dask/issues/7861

**Comment by gmagogsfm** [comment]
Thanks for reporting this issue.

I think your suggestion makes sense. Given that complexity of this task is not high, I will leave it as a task for OSS contribution.

Here are some pointers:

1. A scripted function is a `ScriptFunction` object defined in: https://github.com/pytorch/pytorch/blob/05c1e5b65571d2fa14b0f06c37ba970fd7cc43d0/torch/jit/_script.py#L52
2. We can add a `__getstate__` method to `ScriptFunction`. The method should do nothing more than throwing an exception similar to the one here: https://github.com/pytorch/pytorch/blob/05c1e5b65571d2fa14b0f06c37ba970fd7cc43d0/torch/jit/_script.py#L436
3. A test should also be added to cover new exception

**Comment by SiddeshSambasivam** [comment]
Hi @gmagogsfm ! I'm more than happy to help with this task.

**Comment by gmagogsfm** [comment]
> Hi @gmagogsfm ! I'm more than happy to help with this task.

Awesome! Please see my previous comment for pointers on implementing this improvement. Feel free to ask here if you see any unexpected issue.

**Comment by stubbelj** [comment]
Hi @gmagogsfm, myself and two others are interested in working on this issue as our first open source code contribution. We came into contact with SiddeshSambasivam, who said that it was fine if we took over for them and built on their pull request. Would that be okay?

**Comment by gmagogsfm** [comment]
> Hi @gmagogsfm, myself and two others are interested in working on this issue as our first open source code contribution. We came into contact with SiddeshSambasivam, who said that it was fine if we took over for them and built on their pull request. Would that be okay?

Yeah, totally! Thanks for contributing! Feel free to post any questions you have here and tag me.

**Comment by sayantan1410** [comment]
Hello,
Is the issue unassigned, if so I would like to contribute to it.
Thank you !!

**Comment by itsmingjie** [comment]
Hey @suo — is this issue fixed? I think progress has been made #67076 but a bot just closed the PR. Can you check it out? Thanks.

**Comment by LawJarp-A** [comment]
Hey, is this issue open?
I would like to contribute :)

**Comment by samhu1** [comment]
Is this issue still open?

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
